### PR TITLE
feat(pacmak): Add support for non-Maven Central repository

### PR DESCRIPTION
**Pull request by icj217**
Add CLI arguments that will inject appropriate values into maven settings file so that non-Maven Central dependencies can be accessed at build time

This was specifically written to work with AWS CodeArtifact per the [setup guide](https://docs.aws.amazon.com/codeartifact/latest/ug/maven-mvn.html#fetching-dependencies), but other repositories will work as well.

Example usage: 
```
> jsii-pacmak \
  --target java \
  --maven-extra-repository-id codeartifact \
  --maven-extra-repository-url "https://mydomain-123456789012.d.codeartifact.us-east-1.amazonaws.com/maven/myrepo/"\ 
  --maven-extra-repository-username aws \
  --maven-extra-repository-password "${CODEARTIFACT_AUTH_TOKEN}" 
```

Username/password can be omitted, but ID + URL are required for the maven settings to be updated.

Should also address issue #2404 / discussion #3465 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

**Comment by icj217** [comment]
@RomainMuller Do mind approving the workflow runs and possibly reviewing when you get a chance?

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 1]
```diff
@@ -285,6 +285,43 @@ export class JavaBuilder implements TargetBuilder {
...
+    const extraRepositoryUsername: string | undefined =
+      this.options.arguments['maven-extra-repository-username'];
+    const extraRepositoryPassword: string | undefined =
+      this.options.arguments['maven-extra-repository-password'];
+
+    if (extraRepositoryId) {
+      if (!extraRepositoryUrl)
+        throw new Error(
+          'Extra repository requested but no URL was provided! Use the --maven-extra-repository-url argument to provide a value.',
+        );
```
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/java.ts:311]: Could this maybe be made a single option? Something like `--maven-extra-repository=id=url` or if you need credentials (password caveat from above still applies) `--maven-extra-repository=id=username:password@url` (although maybe `yargs` won't like this format?).

Then - is there a use-case for having more than 1 extra repository configured?

File: packages/jsii-pacmak/bin/jsii-pacmak.ts [hunk 2]
```diff
@@ -153,6 +153,34 @@ import { VERSION_DESC } from '../lib/version';
...
+      default: undefined,
+      hidden: true,
+    })
+    .option('maven-extra-repository-password', {
+      type: 'string',
+      desc: 'Password for authenticating to extra repository',
+      defaultDescription: 'No extra repository is registered.',
+      default: undefined,
+      hidden: true,
+    })
```
Comment (RomainMuller) [inline packages/jsii-pacmak/bin/jsii-pacmak.ts:183]: Passing passwords over CLI arguments is notoriously risky... There's many avenues through which program arguments may be visible by other parties...

I get that these _may_ be "short-lived" credentials here (so less risky to _maybe_ leak somewhere), but I'd still rather have an alternate (safer?) way to provide the password... Not sure what the typical maven practice there is...

This ultimately makes me wonder if we shouldn't instead allow one to provide a "base" XML maven settings file to use, which can include some repos, credentials, etc... This is just an idea, though... I don't know that Maven settings XML documents allow "extending" others without actually copying + modifying them... Which sounds like a lot of unneeded work.

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 3]
```diff
@@ -285,6 +285,43 @@ export class JavaBuilder implements TargetBuilder {
 
     const profileName = 'local-jsii-modules';
     const localRepository = this.options.arguments['maven-local-repository'];
+
+    const repos = localRepos.map((repo) => ({
+      id: repo.replace(/[\\/:"<>|?*]/g, '$'),
+      url: `file://${repo}`,
+    }));
```
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/java.ts:292]: We could simplify this a bit I feel...

```suggestion
    const repos = localRepos.map((repo, idx) => ({
      id: `_local_${idx}`,
      url: `file://${repo}`,
    }));
```

If doing this, we can ensure "no collisions" by documenting & encoding that "extra" repo IDs may not have a name starting with `_`.

File: packages/jsii-pacmak/lib/targets/java.ts [hunk 4]
```diff
@@ -285,6 +285,43 @@ export class JavaBuilder implements TargetBuilder {
...
+    const extraRepositoryPassword: string | undefined =
+      this.options.arguments['maven-extra-repository-password'];
+
+    if (extraRepositoryId) {
+      if (!extraRepositoryUrl)
+        throw new Error(
+          'Extra repository requested but no URL was provided! Use the --maven-extra-repository-url argument to provide a value.',
+        );
+
+      repos.push({ id: extraRepositoryId, url: extraRepositoryUrl });
```
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/java.ts:313]: I suggest turning `repos` into a `Map<string, string>` (key is the ID, value is the URL), so it's easier to guarantee no 2 repositories have the same ID?

**Review by RomainMuller — COMMENTED** [review]
That looks pretty good... I have some suggestions for improvements, but some I'm not even sure myself whether they are actionable in reality...

It would also be nice if you could document usage in https://github.com/aws/jsii/blob/main/gh-pages/content/user-guides/lib-author/configuration/targets/java.md as well.

**Comment by huyphan** [comment]
Instead of exposing every Maven option via the CLI, I suggest that we provide a single option for users to override the setting file that is [currently hard-coded by `jsii-pacmak`](https://github.com/aws/jsii/pull/3893). IMO it's a more future-proof approach and allows users to have more control.

**Comment by RomainMuller** [comment]
> Instead of exposing every Maven option via the CLI, I suggest that we provide a single option for users to override the setting file that is [currently hard-coded by `jsii-pacmak`](https://github.com/aws/jsii/blob/d2ecb6d2a6a460e309e4985dc001a6b749b3aac7/packages/jsii-pacmak/lib/targets/java.ts#L450). IMO it's a more future-proof approach and allows users to have more control.

I agree that this might offer more flexibility...

**Comment by mrgrain** [comment]
Solved by https://github.com/aws/jsii/pull/4068

**Pull request closed without merging**